### PR TITLE
Galactic Radio Telescope Update

### PR DESCRIPTION
**Pull request by hexylena**
replaces #2768. Many changes and optimisations:

- query tables directly with sqlalchemy rather than allowing it to decode into objects
- optional sanitisation (based on original work by @ericenns) at the tool level and at the parameter level (slow).
- separate reporting + uploading steps to allow admins behind firewalls to help out with the GRT projcet.
- remove instance metadata nonsense since other projects want to handle that and I don't care about it. I just want your jobs :)
- export performance is on the order of 150k jobs / minute (for admin reference)

Updated backend is available, but it is quite boring. Just accepts reports, and then processes them offline (on cron), and generates some reports.

### CODE REVIEW

File: scripts/grt-submit.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
```
Comment (bgruening) [inline scripts/grt-submit.py]: We should but this into it's own subdir.

File: scripts/grt-submit.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('-r', '--report-directory', help='Directory in which reports are stored',
+                        default=os.path.abspath(os.path.join('.', 'reports')))
+    parser.add_argument('-c', '--config', help='Path to GRT config file',
+                        default=default_config)
+    parser.add_argument("-l", "--loglevel", choices=['debug', 'info', 'warning', 'error', 'critical'],
+                        help="Set the logging level", default='warning')
+    args = parser.parse_args()
+
+    logging.info('Loading GRT ini...')
```
Comment (bgruening) [inline scripts/grt-submit.py]: ini? config?

File: scripts/grt-submit.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    try:
+        with open(args.config) as handle:
+            config = yaml.load(handle)
+    except Exception:
+        logging.info('Using default GRT Configuration')
```
Comment (bgruening) [inline scripts/grt-submit.py]: s/C/c/

File: scripts/grt-submit.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    r = requests.post(GRT_URL + 'api/whoami', headers=headers)
+    data = r.json()
+    # we get back some information about which reports had previously been uploaded.
+    remote_reports = data['uploaded_reports']
+    # so now we can know which to send.
+    local_reports = [x.strip('.json') for x in os.listdir(REPORT_DIR) if x.endswith('.json')]
+    for report_id in local_reports:
+        if report_id not in remote_reports:
+            print("Uploading %s" % report_id)
```
Comment (bgruening) [inline scripts/grt-submit.py]: logging.info

File: scripts/grt-submit.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            files = {
+                'meta': open(os.path.join(sys.argv[1], report_id + '.json'), 'rb'),
+                'data': open(os.path.join(sys.argv[1], report_id + '.tsv.gz'), 'rb')
+            }
+            data = {
+                'identifier': report_id
+            }
+            r = requests.post(GRT_URL + 'api/v2/upload', files=files, headers=headers, data=data)
+            print(r.json())
```
Comment (bgruening) [inline scripts/grt-submit.py]: This line should be removed, I guess.

File: scripts/grt.py [hunk 6]
```diff
@@ -43,184 +56,325 @@ def _init(config):
...
-    if args.instance_id:
-        config_dict['instance_id'] = args.instance_id
-    if args.api_key:
-        config_dict['api_key'] = args.api_key
-    if args.grt_url:
-        config_dict['grt_url'] = args.grt_url
-
-    print('Loading Galaxy...')
-    model, object_store, engine = _init(config_dict['galaxy_config'])
+        logging.info('Using default GRT Configuration')
```
Comment (bgruening) [inline scripts/grt.py]: s/Configuration/configuration/

File: scripts/grt.py [hunk 7]
```diff
@@ -1,22 +1,28 @@
...
+import json
+import os
 import sqlalchemy as sa
+import subprocess
+import sys
+import time
 import yaml
-import re
+import logging
+# logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)
```
Comment (bgruening) [inline scripts/grt.py]: This can be removed.

File: scripts/grt.py [hunk 8]
```diff
@@ -43,184 +56,325 @@ def _init(config):
...
+    handle_job = open(REPORT_BASE + '.jobs.tsv', 'w')
+    handle_job.write('\t'.join(('id', 'tool_id', 'tool_version', 'state', 'create_time')) + '\n')
+    for offset_start in range(last_job_sent, end_job_id, args.batch_size):
+        logging.debug("Processing %s:%s", offset_start, min(end_job_id, offset_start + args.batch_size))
+        for job in sa_session.query(model.Job.id, model.Job.user_id, model.Job.tool_id, model.Job.tool_version, model.Job.state, model.Job.create_time) \
+                .filter(model.Job.id > offset_start) \
+                .filter(model.Job.id <= min(end_job_id, offset_start + args.batch_size)) \
+                .all():
+
+            # TODO: blacklisted
```
Comment (bgruening) [inline scripts/grt.py]: Can you write a little bit more about the TODO here.

**Review by bgruening — COMMENTED** [review]
Thanks @erasche! Some small comments inline.

File: scripts/grt-submit.py (hunk 1 above)
Comment (hexylena) [reply scripts/grt-submit.py]: sure, I can do this. I guess scripts is getting a bit full.

**Comment by hexylena** [comment]
Addressed.

**Comment by hexylena** [comment]
There is currently some small problem with "share_toolbox" so I've defaulted it to false. Having the full toolbox is less important but still a useful thing to collect since we can produce useful reports for admins. I will patch this in a future PR, but it is not a blocking issue for this one I believe.

**Comment by jmchilton** [comment]
Thanks for the update and continued work on this @erasche - I look forward to great things!

One thing I'd like to see is that if we are claiming to block certain parameters that we provide a test case to verify this blocking is in fact happening and doesn't get broken at some point. Especially given the complexity of nested parameters and such. Just something to think about for future development.

Thanks again though!

**Comment by hexylena** [comment]
@jmchilton where would be a good place to do that? Should I move sanitization into lib/galaxy/ somewhere and then provide a test case?

**Comment by jmchilton** [comment]
I think you should be able to setup an API test that runs a bunch of jobs and then calls the script to generate the data dump directly from the API test - use the path to the test case Python file to find galaxy_root and then append the scripts directory - call Python directly against that script after setting up test files to target the test setup. I'm not sure things need to be refactored. It might be more complicated than that - I don't know - but it seems straight forward at first glance. Integration tests allow more customization of Galaxy's config prior to starting up Galaxy - so if it is more complicated than an API test allows and you needed to target that directly you could think about doing it there.

**Pull request merged**